clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -1618,20 +1642,60 @@ recToLetRec (TransformContext is0 []) e = do
 
     eqArg _ v1 v2@(Var {})
       = v1 == v2
-    eqArg tcm v1 v2@(collectArgs -> (Data _,args'))
+    eqArg tcm v1 v2@(collectArgs -> (Data _, args'))
       | termType tcm v1 == termType tcm v2
-      = and (zipWith (isNthProjection v1) [0..] (Either.lefts args'))
+      = if isClassConstraint (termType tcm v1) then
+          -- Class constraints are equal if their types are equal, so we can
+          -- take a shortcut here.
+          True
+        else
+          -- Check whether all arguments to the data constructor are projections
+          --
+          and (zipWith (eqDat v1) (map pure [0..]) (Either.lefts args'))
     eqArg _ _ _
       = False
 
-    -- `isNthProjection s n c` checks that `c` is the `n`th projection
-    -- of `s`.
-    isNthProjection :: Term -> Int -> Term -> Bool
-    isNthProjection v n (Case v' _ [(DataPat _ _ xs, Var s)])
-      | v == v'
-      , Just n' <- List.elemIndex s xs
-      = n == n'
-    isNthProjection _ _ _ = False
+    -- Recursively check if data construction is equal to some variable.
```

Comment:
```whether
-- Recursively check whether a term /e/ is semantically equal to some variable /v/.
-- Currently it can only assert equality when /e/ is  syntactically equal
-- to /v/, or is constructed out of projections of /v/, importantly:
--
-- [Note: Breaks on constants and predermined equality]
-- ...
```